### PR TITLE
feat: APKAM sync changes

### LINKED ISSUE #1472: Issues in sync with APKAM changes
### Describe the bug

- [ ] Wavi app keys are created and synced to the cloud
- [ ] Connected to another client which has enrolled to buzz namespace and Tried to run sync, by using the command sync:from. All the keys with wavi namespace are also synced
- [x] Scan verb displays each key twice -- This issue is not reproducible with the latest changes

### Steps to reproduce

Attaching the test steps in the ticket

[test_logs_apkam_buzz_namespace.txt](https://github.com/atsign-foundation/at_server/files/12146639/test_logs_apkam_buzz_namespace.txt)

### Expected behavior

Only buzz keys should be synced

### Screenshots

_No response_

### Were you using an atApplication when the bug was found?

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sitaram-kalluri**
**- What I did**
- When connection is authenticated via the APKAM
   - Enhance the sync verb to send the keys associated with enrolled namespaces in the sync response for the client to sync.
   - Enhance the "stats" verb to provide the highest commitId among the enrolled namespaces.

**- How I did it**
- Within the `SyncProgressiveVerbHandler`,`prepareResponse` method, when the connection is authenticated through APKAM, retrieve the enrolled namespaces. If the key's namespace is found among the enrolled namespaces or if the enrolled namespaces contain "*", include the key entry in the sync response for client to sync.

- With the above changes, the following scenario causes the sync to run into infinite loop : If a client submits a regex containing two namespaces, where one of the namespaces lacks authorization, and the key with the unauthorized namespace possesses the highest commitId, it can potentially lead to an infinite loop within the client's sync service. This issue arises because the server fails to send the commit entry with the highest commitId. To address the problem, enhance the "stats:3" verb to send the latestCommitId among the enrolled namespaces.

- The stats_verb_handler, call the "getMetrics" method to fetch the latestCommitId. In the `getMetrics` method within the `LastCommitIDMetricImpl`, introduce an optional parameter called `enrolledNamespaces` that is set within the `stats_verb_handler` when the connection is authenticated via APKAM.

- Modify the `getMetrics` method to pass the `enrolledNamespaces` parameter to the `atCommitLog->lastCommittedSequenceNumberWithRegex` method in order to filter the keys based on the enrolled namespaces.

- In the `CommitLogKeyStore.dart` file, include a new method named `isNamespaceAuthorized` to validate whether the key's namespace is authorized.

**- How to verify it**
- Added unit tests in "stats_verb_test.dart" and "sync_verb_test.dart"
- Added functional test in "enroll_verb_test.dart" to verify highest commit id among the enrolled namespaces is returned
- Added functional test in "sync_verb_test.dart" to verify keys with enrolled namespaces are returned in the sync response

**- Description for the changelog**
- Implement APKAM changes into sync verb

NOTE: 
1. Added a work around https://github.com/atsign-foundation/at_server/pull/1569/commits/e9647ae2b6c19171e9746e121493d99e5b66b3fd for the bug: https://github.com/atsign-foundation/at_server/issues/1570 in "commit_log_keystore.dart"
2. The changes in at_persistence_secondary_server which are a part of this PR are also in "apkam_sync_persistence_changes" which has CHANGELOG and pubspec version updated.

### CODE REVIEW

**Review by gkc — COMMENTED** [review]
@sitaram-kalluri Have you run the at_client tests against this version of the server? (i.e. local test runs)

**Comment by sitaram-kalluri** [comment]
> @sitaram-kalluri Have you run the at_client tests against this version of the server? (i.e. local test runs)

@gkc: Ran at_client functional tests pointing this version of secondary server and all tests passed.

File: packages/at_secondary_server/lib/src/verb/handler/stats_verb_handler.dart [hunk 1]
```diff
@@ -108,12 +111,21 @@ class StatsVerbHandler extends AbstractVerbHandler {
...
-    if ((id == '3' || id == '15') && _regex != null) {
+    if (id == '3') {
+      if (_regex == null || _regex.isEmpty) {
+        _regex = '.*';
+      }
+      value = await (metric.name as LastCommitIDMetricImpl)
+          .getMetrics(regex: _regex, enrolledNamespaces: enrolledNamespaces);
+    } else if (id == '15' && _regex != null) {
+      // When connection is authenticated via the APKAM, return the highest commit-Id
+      // among the specified namespaces.
```
Comment (gkc) [inline packages/at_secondary_server/lib/src/verb/handler/stats_verb_handler.dart]: Shouldn't these two comment lines should be in the `id == '3'` block above?

File: packages/at_secondary_server/pubspec.yaml [hunk 2]
```diff
@@ -35,6 +35,14 @@ dependencies:
   yaml: 3.1.2
   logging: 1.2.0
 
+dependency_overrides:
+  at_persistence_secondary_server:
+    git:
+      url: https://github.com/atsign-foundation/at_server.git
+      path: packages/at_persistence_secondary_server
+      ref: apkam_sync_persistence_changes
```
Comment (gkc) [inline packages/at_secondary_server/pubspec.yaml]: Please remove dependency override

I note that the `apkam_sync_persistence_changes` branch looks a bit stale - if it is not an active branch then it should be deleted.

File: packages/at_secondary_server/test/stats_verb_test.dart [hunk 3]
```diff
@@ -124,7 +124,7 @@ void main() {
   group('A group of notificationStats verb tests', () {
     SecondaryKeyStoreManager? keyStoreManager;
     setUp(() async => keyStoreManager = await setUpFunc(
-        '${Directory.current.path}/test/hive',
+        '${Directory.current.path}/unit_test_storage',
```
Comment (gkc) [inline packages/at_secondary_server/test/stats_verb_test.dart:127]: why the change? (just curious)

File: tests/at_functional_test/pubspec.yaml [hunk 4]
```diff
@@ -20,6 +20,13 @@ dependencies:
   uuid: ^3.0.7
   elliptic: ^0.3.8
 
+dependency_overrides:
+  at_persistence_secondary_server:
+    git:
+      url: https://github.com/atsign-foundation/at_server.git
+      path: packages/at_persistence_secondary_server
+      ref: apkam_sync_persistence_changes
+
```
Comment (gkc) [inline tests/at_functional_test/pubspec.yaml]: remove dependency override

**Review by gkc — CHANGES_REQUESTED** [review]
LGTM overall but needs some tidying up before we can merge

File: packages/at_secondary_server/test/stats_verb_test.dart (hunk 3 above)
Comment (sitaram-kalluri) [reply packages/at_secondary_server/test/stats_verb_test.dart:127]: At the end of the group added a tearDown method from test_utils.dart - `tearDown(() async => await verbTestsTearDown());` which clears from "unit_test_storage". Hence marked the storage directory to "unit_test_storage" to ensure the test data is cleared at the end of the test.

File: tests/at_functional_test/pubspec.yaml (hunk 4 above)
Comment (sitaram-kalluri) [reply tests/at_functional_test/pubspec.yaml]: After merging the PR: https://github.com/atsign-foundation/at_server/pull/1619, will replace the dependency_overrides with hosted version.

File: packages/at_secondary_server/lib/src/verb/handler/stats_verb_handler.dart (hunk 1 above)
Comment (sitaram-kalluri) [reply packages/at_secondary_server/lib/src/verb/handler/stats_verb_handler.dart]: Moved the code comments above.

File: packages/at_secondary_server/pubspec.yaml (hunk 2 above)
Comment (sitaram-kalluri) [reply packages/at_secondary_server/pubspec.yaml]: Removed dependency overrides.

**Review by gkc — APPROVED** [review]
LGTM - please make canary release c3.0.36b once this PR has been merged
